src/ORM/Related/DataObjectRelatedService.php

```diff
@@ -0,0 +1,332 @@
+<?php
+
+namespace SilverStripe\ORM\Related;
+
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Config;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\Connect\Query;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DataObjectSchema;
+use SilverStripe\ORM\Queries\SQLSelect;
+
+class DataObjectRelatedService
+{
+
+    /**
+     * Used to prevent duplicate database queries
+     *
+     * @var array
+     */
+    private $queryIdens = [];
+
+    /**
+     * Find all DataObject instances that have a linked relationship with $record
+     *
+     * @param DataObject $record
+     * @param array $excludedClasses
+     * @return ArrayList
+     * @throws \ReflectionException
+     */
+    public function findAll(DataObject $record, array $excludedClasses = []): ArrayList
```

Comment:
```suggestion
    public function findAll(DataObject $record, array $excludedClasses = []): SSList
```

Reply:
I still think there's a lot of opportunity to split this method into smaller more manageable chunks.

Reply:
I refactored the whole thing, lots of private methods now